src/TorchSharp/DataLoader.cs

```diff
@@ -111,17 +121,50 @@ public bool MoveNext()
                         {
                             DisposeCurrent();
                             if (!MoveNextValue()) return false;
-                            List<Dictionary<string, Tensor>> dic = new();
-                            dic.Add(dataset.GetTensor(currentVal));
-                            for (var i = 1; i < batchSize; i++) {
+
+                            var tensorIndexList = new List<long> {currentVal};
+                            for (int i = 1; i < batchSize; i++) {
                                 if (!MoveNextValue()) break;
-                                dic.Add(dataset.GetTensor(currentVal));
+                                tensorIndexList.Add(currentVal);
                             }
 
+                            if (!shuffle)
+                                tensorIndexList.Reverse(); // ConcurrentBag::Add inserts data into first of the collections. So, I'll reverse them.
+
+                            ConcurrentBag<Dictionary<string, Tensor>> dic = new();
+                            var taskedBatchCount = 0;
+                            var taskBatchLock = new object();
+
+                            //Run Async
+                            foreach(var _ in Enumerable.Range(1, num_worker - 1))
+                                ThreadPool.QueueUserWorkItem(CreateBatch);
```

Comment:
I would suggest using the more modern Task API, instead of the thread pool. That also allows you to wait more efficiently (not spin in a loop).

See: 

https://dotnetcodr.com/2014/01/01/5-ways-to-start-a-task-in-net-c

and:

https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task?view=net-6.0

That allows you to collect all the task instances and wait for them with one call. You can still use the CreateBatch() that you have.